File: Languages/python/algorithms-dsa/leetcode/trees/level_order_traversal.py
"""
LeetCode #102 - Binary Tree Level Order Traversal
Topic: Trees
Difficulty: Medium

BFS level order traversal using deque.

Time Complexity: O(n)
Space Complexity: O(n)
"""
from collections import deque


class TreeNode:
    def __init__(self, val: int = 0, left: "TreeNode | None" = None, right: "TreeNode | None" = None):
        self.val = val
        self.left = left
        self.right = right


def level_order(root: TreeNode | None) -> list[list[int]]:
    if not root:
        return []
    result: list[list[int]] = []
    queue = deque([root])
    while queue:
        level: list[int] = []
        for _ in range(len(queue)):
            node = queue.popleft()
            level.append(node.val)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)
        result.append(level)
    return result


if __name__ == "__main__":
    tree = TreeNode(3, TreeNode(9), TreeNode(20, TreeNode(15), TreeNode(7)))
    assert level_order(tree) == [[3], [9, 20], [15, 7]]
    assert level_order(TreeNode(1)) == [[1]]
    assert level_order(None) == []
    print("All tests passed!")

File: Languages/cpp/algorithms-dsa/leetcode/trees/level_order_traversal.cpp
/**
 * LeetCode 102: Binary Tree Level Order Traversal
 * Topic: Trees
 * Difficulty: Medium
 *
 * Return the level order traversal of a binary tree (BFS with queue).
 * Time: O(n), Space: O(n)
 */
#include <iostream>
#include <vector>
#include <queue>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

vector<vector<int>> levelOrder(TreeNode* root) {
    vector<vector<int>> res;
    if (!root) return res;
    queue<TreeNode*> q;
    q.push(root);
    while (!q.empty()) {
        int sz = q.size();
        vector<int> level;
        for (int i = 0; i < sz; i++) {
            TreeNode* node = q.front(); q.pop();
            level.push_back(node->val);
            if (node->left) q.push(node->left);
            if (node->right) q.push(node->right);
        }
        res.push_back(level);
    }
    return res;
}

int main() {
    // Tree: [3, 9, 20, null, null, 15, 7]
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(9);
    root->right = new TreeNode(20);
    root->right->left = new TreeNode(15);
    root->right->right = new TreeNode(7);

    auto res = levelOrder(root);
    assert(res.size() == 3);
    assert(res[0] == vector<int>({3}));
    assert(res[1] == vector<int>({9, 20}));
    assert(res[2] == vector<int>({15, 7}));

    // Empty tree
    assert(levelOrder(nullptr).empty());

    cout << "All tests passed!" << endl;
    return 0;
}
